### ISSUE TITLE
Report error on non-nullable field in class with redirecting factory constructor

### ISSUE THREAD

**Issue by johnniwinther** [dart-lang/sdk#42844]
CFE is current not reporting an error for non-nullable fields in classes with only a factory constructor. 

Multiple tests in `co19/LanguageFeatures/nnbd/static_errors_A06_t*` with the following pattern are currently failing:
```Dart
class C {
  Never n; // Expected error here.

  factory C.f() = D;
}

class D implements C {
  Never n;

  D(this.n);
}
```

**Comment by johnniwinther** [comment]
It seems inconsistent that we report an error in case but not in similar cases of a final fields:
```
class C {
  final dynamic n; // No error here.

  factory C.f() = D;
}

class D implements C {
  final var n;

  D(this.n);
}```

@eernstg @leafpetersen Shouldn't we emit an error in both cases or none?

**Comment by eernstg** [comment]
The example should be adjusted to match up the parameter lists, and then I'll add a mixin application. With that in place we can see that there must be  an error:

```dart
class C {
  Never n; // Expected error here.
  factory C(Never n) = D;
}

class D implements C {
  Never n;
  D(this.n);
}

class E with C {}

void main() {
  var b = false;
  if (b) C(throw 1);
  print(E().n); // Prints 'null'.
}
```

The problem that creates the need for an error is that `C` can be used to derive a mixin, as shown in class `E`, and that allows us to see the uninitialized value of `n` (which happens to be null, which is a soundness violation).

The variant using a final variable has the same soundness issue:

```dart
class A {
  final Never n;
  factory A() = B;
}

class B implements A {
  Never get n => throw 1;
}

class C = Object with A;

void main() {
  print(C().n); // Prints 'null'.
}
```

So we must have a compile-time error in both cases.

The analyzer reports the error with the final instance variable already, so it is a non-breaking change to introduce the error in the CFE as well.

**Comment by eernstg** [comment]
Turns out that the wording in the language specification does _not_ make the final instance variable example an error, but that is surely an oversight. Created https://github.com/dart-lang/language/pull/1130 in order to fix that.

**Comment by leafpetersen** [comment]
I don't believe the analyzer currently issues an error on this, so it would be breaking to change this in pre null-safety code, if that is what is being discussed.

Post-null safety, I believe the already specified errors are sufficient for soundness, but we may wish to add an additional error for the nullable case as well?

**Comment by eernstg** [comment]
Right &mdash; further comments in https://github.com/dart-lang/language/pull/1130#issuecomment-669160076.

But it is still true that the CFE should report the error for the potentially non-nullable case. This issue is still relevant for handling that.

The anomaly about final instance variables with a nullable type is discussed in dart-lang/language#1130.

**Comment by johnniwinther** [comment]
@stefantsov This might be fixed in conjunction with #42967 and #40951

**Comment by stefantsov** [comment]
Resolved by https://github.com/dart-lang/sdk/commit/734da2ca2cd0119140a00340ee2ffe47372fc956.
